🐛 Bug: Public `AnyRule` type is contravariant on `OptionsSchema`

### Bug Report Checklist

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have pulled the latest `main` branch of the repository.
- [x] I have [searched for related issues](https://github.com/JoshuaKGoldberg/flint/issues?q=is%3Aissue) and found none that matched my issue.

### Expected

Expected the following snippet to work:

```ts
import { flint, ts } from 'flint'

defineConfig({
	use: [
		{
			rules: ts.rules({ anyReturns: true })
		}
	]
})
```

### Actual

Running `tsc --noEmit` reports:

```
flint.config.ts:8:4 - error TS2322: Type '[Rule<{ readonly description: "Reports returning a value with type `any` from a function."; readonly id: "anyReturns"; readonly preset: "logical"; }, TSNodesByName, TypeScriptServices, "unsafeReturn" | ... 1 more ... | "unsafeReturnThis", undefined>, ... 48 more ..., Rule<...>]' is not assignable to type 'AnyLevelDeep<ConfigRuleDefinition> | undefined'.
  Type '[Rule<{ readonly description: "Reports returning a value with type `any` from a function."; readonly id: "anyReturns"; readonly preset: "logical"; }, TSNodesByName, TypeScriptServices, "unsafeReturn" | ... 1 more ... | "unsafeReturnThis", undefined>, ... 48 more ..., Rule<...>]' is not assignable to type 'AnyLevelDeep<ConfigRuleDefinition>[]'.
    Type 'Rule<{ readonly description: "Reports returning a value with type `any` from a function."; readonly id: "anyReturns"; readonly preset: "logical"; }, TSNodesByName, TypeScriptServices, "unsafeReturn" | ... 1 more ... | "unsafeReturnThis", undefined> | ... 48 more ... | Rule<...>' is not assignable to type 'AnyLevelDeep<ConfigRuleDefinition>'.
      Type 'Rule<{ readonly description: "Reports returning a value with type `any` from a function."; readonly id: "anyReturns"; readonly preset: "logical"; }, TSNodesByName, TypeScriptServices, "unsafeReturn" | ... 1 more ... | "unsafeReturnThis", undefined>' is not assignable to type 'AnyLevelDeep<ConfigRuleDefinition>'.
        Type 'Rule<{ readonly description: "Reports returning a value with type `any` from a function."; readonly id: "anyReturns"; readonly preset: "logical"; }, TSNodesByName, TypeScriptServices, "unsafeReturn" | ... 1 more ... | "unsafeReturnThis", undefined>' is not assignable to type 'AnyRule'.
          Types of property 'setup' are incompatible.
            Type 'RuleSetup<TSNodesByName, TypeScriptServices, "unsafeReturn" | "unsafeReturnAssignment" | "unsafeReturnThis", undefined>' is not assignable to type 'RuleSetup<any, any, string, Partial<Record<string, unknown>> | undefined>'.
              Types of parameters 'options' and 'options' are incompatible.
                Type 'Partial<Record<string, unknown>> | undefined' is not assignable to type 'undefined'.
                  Type 'Partial<Record<string, unknown>>' is not assignable to type 'undefined'.

8    rules: ts.rules({ anyReturns: true })
     ~~~~~

  packages/core/lib/types/configs.d.ts:21:5
    21     rules?: AnyLevelDeep<ConfigRuleDefinition>;
           ~~~~~
    The expected type comes from property 'rules' which is declared here on type 'ConfigUseDefinition'
```

### Additional Info

This is because `AnyRule` passes `OptionsSchema = Partial<Record<string, unknown>> | undefined` to `RuleDefinition`:

https://github.com/JoshuaKGoldberg/flint/blob/14abad1c2b21ad79baf447d78e0eaa9834ebe4cb/packages/core/src/types/rules.ts#L10-L12

Meanwhile, `RuleDefinition` uses this `OptionsSchema` as the `setup` function's argument type. Thus, `RuleDefinition` is contravariant on `OptionsSchema`.

https://github.com/JoshuaKGoldberg/flint/blob/14abad1c2b21ad79baf447d78e0eaa9834ebe4cb/packages/core/src/types/rules.ts#L77-L82

https://github.com/JoshuaKGoldberg/flint/blob/14abad1c2b21ad79baf447d78e0eaa9834ebe4cb/packages/core/src/types/rules.ts#L90-L98

This can be fixed if `AnyRule`'s `setup` function accepts `options` typed as `any`.

--- a/.changeset/petite-feet-dance.md
+++ b/.changeset/petite-feet-dance.md
@@ -1,0 +1,5 @@
+---
+"@flint.fyi/core": patch
+---
+
+fix(core): make `Rule` type assignable to `AnyRule`

--- a/packages/core/src/types/rules.test-d.ts
+++ b/packages/core/src/types/rules.test-d.ts
@@ -1,0 +1,69 @@
+import { describe, expectTypeOf, it } from "vitest";
+import { z } from "zod";
+
+import { AnyRule, AnyRuleDefinition, Rule } from "./rules.js";
+import { AnyOptionalSchema, InferredObject } from "./shapes.js";
+
+describe("Rule", () => {
+	it("should be assignable to AnyRule", () => {
+		const rule = {} as Rule<
+			{
+				description: "desc";
+				id: "id";
+			},
+			{ bar: "bar" },
+			{ bar: "bar" },
+			"message",
+			undefined
+		>;
+
+		const fn = (r: AnyRule) => r;
+
+		fn(rule);
+	});
+
+	it("should be assignable to AnyRuleDefinition", () => {
+		const rule = {} as Rule<
+			{
+				description: "desc";
+				id: "id";
+			},
+			{ bar: "bar" },
+			{ bar: "bar" },
+			"message",
+			undefined
+		>;
+
+		const fn = (r: AnyRuleDefinition) => r;
+
+		fn(rule);
+	});
+
+	it("should propagate options schema in setup", () => {
+		type rule = Rule<
+			{
+				description: "desc";
+				id: "id";
+			},
+			{ bar: "bar" },
+			{ bar: "bar" },
+			"message",
+			{
+				foo: z.ZodOptional<z.ZodString>;
+			}
+		>;
+
+		const options = {} as Parameters<rule["setup"]>[1];
+
+		expectTypeOf(options).toEqualTypeOf<{ foo?: string | undefined }>();
+	});
+
+	it("should work with AnyOptionalSchema|undefined", () => {
+		void (<T extends AnyOptionalSchema | undefined>(
+			rule: AnyRuleDefinition<T>,
+			options: InferredObject<T>,
+		) => {
+			void rule.setup({}, options);
+		});
+	});
+});

--- a/packages/core/src/types/rules.ts
+++ b/packages/core/src/types/rules.ts
@@ -5,38 +5,16 @@ import { PromiseOrSync } from "./promises.js";
 import { ReportMessageData } from "./reports.js";
 import { AnyOptionalSchema, InferredObject } from "./shapes.js";
 
+/* eslint-disable @typescript-eslint/no-explicit-any */
 export type AnyRule<
 	About extends RuleAbout = RuleAbout,
-	OptionsSchema extends AnyOptionalSchema | undefined =
-		| AnyOptionalSchema
-		| undefined,
-> = Rule<
-	About,
-	// TODO: How to make types more permissive around assignability?
-	// eslint-disable-next-line @typescript-eslint/no-explicit-any
-	any,
-	// TODO: How to make types more permissive around assignability?
-	// eslint-disable-next-line @typescript-eslint/no-explicit-any
-	any,
-	string,
-	OptionsSchema
->;
+	OptionsSchema extends AnyOptionalSchema | undefined = any,
+> = Rule<About, any, any, string, OptionsSchema>;
 
 export type AnyRuleDefinition<
-	OptionsSchema extends AnyOptionalSchema | undefined =
-		| AnyOptionalSchema
-		| undefined,
-> = RuleDefinition<
-	RuleAbout,
-	// TODO: How to make types more permissive around assignability?
-	// eslint-disable-next-line @typescript-eslint/no-explicit-any
-	any,
-	// TODO: How to make types more permissive around assignability?
-	// eslint-disable-next-line @typescript-eslint/no-explicit-any
-	any,
-	string,
-	OptionsSchema
->;
+	OptionsSchema extends AnyOptionalSchema | undefined = any,
+> = RuleDefinition<RuleAbout, any, any, string, OptionsSchema>;
+/* eslint-enable @typescript-eslint/no-explicit-any */
 
 /**
  * A single lint rule, as used by users in configs.

--- a/packages/core/src/types/shapes.ts
+++ b/packages/core/src/types/shapes.ts
@@ -22,6 +22,9 @@ export type AnyOptionalSchema = Record<
  */
 export type InferredObject<
 	OptionsSchema extends AnyOptionalSchema | undefined,
-> = OptionsSchema extends AnyOptionalSchema
-	? Partial<z.infer<z.ZodObject<OptionsSchema>>>
-	: undefined;
+> = 0 extends 1 & NoInfer<OptionsSchema>
+	? // propagate `any`
+		OptionsSchema
+	: OptionsSchema extends AnyOptionalSchema
+		? Partial<z.infer<z.ZodObject<OptionsSchema>>>
+		: undefined;

